lib/html_cleaner.rb

```diff
@@ -290,20 +294,17 @@ def traverse_nodes(node, stack=nil, out_html=nil)
     end
 
     if !node.text? && !node.cdata?
+      out_html += stack.close_and_pop_last(1) if put_outside_p_tag?(node.name) && stack.immediately_inside_paragraph?
+
       out_html += stack.add_p if put_inside_p_tag?(node.name) && !stack.inside_paragraph?
 
       stack << [[node.name, Hash[*(node.attribute_nodes.map { |n| [n.name, n.value] }.flatten)]]]
       out_html += open_tag(node)
 
       # If we are the root node, pre-emptively open a paragraph
-      if node.name == "myroot"
-        out_html += stack.add_p
-      end
-
-      if no_break_before_after_tag?(node.name) and !stack.last.include?("p")
-        out_html += stack.add_p
-      end
+      out_html += stack.add_p if node.name == "myroot" && node.children.any? { |x| put_inside_p_tag?(x.name) || x.text? || x.cdata? }
```

Comment:
I believe the condition ` node.name == "myroot"` is too loose and adds unnecessary `p` tags a lot.

However, the additional conditions may degrade performance.